src/tests/Microsoft.PowerFx.Interpreter.Tests/CustomFunctions.cs

```diff
@@ -38,6 +42,82 @@ public static StringValue Execute(NumberValue x, BooleanValue b)
             }
         }
 
+        [Fact]
+        public async void SimpleCustomAsyncFuntion()
+        {
+            var config = new PowerFxConfig(null);
+
+            config.AddFunction(new TestCustomAsyncFunction());
+            var engine = new RecalcEngine(config);
+
+            // Shows up in enumeration
+            var func = engine.GetAllFunctionNames().First(name => name == "TestCustomAsync");
+            Assert.NotNull(func);
+
+            // Can be invoked. 
+            using var cts = new CancellationTokenSource();
+            var resultAsync = await engine.EvalAsync("TestCustomAsync(3, true)", cts.Token);
+
+            Assert.Equal("3,True", resultAsync.ToObject());
+        }
+
+        // Verify a custom function can return a non-completed task. 
+        [Fact]
+        public async Task VerifyCustomFunctionIsAsync()
+        {
+            var func = new TestCustomWaitAsyncFunction();
+            var config = new PowerFxConfig(null);
+            config.AddFunction(func);
+
+            var engine = new RecalcEngine(config);
+
+            // Can be invoked. 
+            using var cts = new CancellationTokenSource();
+
+            var task = engine.EvalAsync("TestCustomWaitAsync()", cts.Token);
+            await Task.Yield();
+
+            // custom func is blocking on our waiter
+            await Task.Delay(TimeSpan.FromMilliseconds(5));
+            Assert.False(task.IsCompleted);
+
+            func.SetResult(15);
+
+            var result = await task;
+
+            Assert.Equal(30.0, result.ToObject());
+        }
+
+        private class TestCustomAsyncFunction : ReflectionFunction
+        {
+            public static async Task<StringValue> Execute(NumberValue x, BooleanValue b)
+            {
+                var val = x.Value.ToString() + "," + b.Value.ToString();
+                return FormulaValue.New(val);
+            }
+        }
+
+        private class TestCustomWaitAsyncFunction : ReflectionFunction
+        {
+            private readonly TaskCompletionSource<FormulaValue> _waiter = new TaskCompletionSource<FormulaValue>();
+
+            public async Task<NumberValue> Execute()
```

Comment:
another test: async functions should be able to model bind to the CancellationToken.